labs/LanguageBasics/LanguageBasics.md

```diff
@@ -43,55 +285,84 @@ labs::LanguageBasics::LanguageBasics> gcdCurried 20 28
+  ```sh
+  labs::LanguageBasics::LanguageBasics> :type gcdCurried 10
+  gcdCurried 10 : Integer -> Integer
+  ```
+  `gcdCurried 10` takes an integer and returns an integer. When it
+  is applied to an integer it computes the gcd of 10 and that
+  integer. Other examples to illustrate partial application:
+  * Incrementing is addition partially applied to 1. Notionally:
+    `inc x = (add 1) x`
+  * The reciprocal is division partially applied to 1. Notionally:
+    `recip x = (div 1) x`
+* Really `gcdUncurried` is a function of one argument. That argument
+  is an ordered pair which makes `gcdUncurried (28, 20)` look just
+  like a two argument function in many languages.
+
+
+Small Functions
+---------------
+
+Cryptol programs are just sequences of appropriate functions applied
+in the correct order. Good Cryptol features small, easy to understand
+functions composed into conceptually bigger ones. This is good
+computer science in general, but in Cryptol it is even more
+advantageous:
+* Easy to test—Cryptol's interpreter makes it very cheap to try your
+  functions out.
+* Encourages programming with properties—Properties can be tested
+  easily and, as we'll see in other labs, proven to provide guarantees
+  about code. Moreover, properties serve as another kind of
+  documentation!
 
 ### Examples
 
 ```
 abs : Integer -> Integer
 abs n = if n >= 0 then n else -n
 
-abs_nonnegative : Integer -> Bit
-property abs_nonnegative x = abs x >= 0
+absNonnegative : Integer -> Bit
+property absNonnegative x = abs x >= 0
 ```
 
 * `abs : Integer -> Integer` is the type signature for `abs`
 * `abs n = if n >= 0 then n else -n` is the definition for `abs` (or function body)
-* `property abs_nonnegative ...` is a property we expect the function to have.
-    * `:check property abs_nonnegative` checks this property with random tests. It's super cheap unit testing!
-    ```sh
-    Main> :check abs_nonnegative 
-    Using random testing.
-    Passed 100 tests.
-    ```
-* Cryptol's `if`-`then`-`else` is much like C's ternary operator `?`...`:`. It is not like the `if`-`then`-`else` control structure.
+* `property absNonnegative ...` is a property we expect the function to have.
+* `:check property absNonnegative` checks this property with
+    random tests. It's super cheap unit testing!
+  ```sh
+  Main> :check absNonnegative 
+  Using random testing.
+  Passed 100 tests.
+  ```
+* Cryptol's `if`/`then`/`else` is much like C's ternary operator
+  `?`...`:`. It is not like the `if`/`then`/`else` control structure.
 * The reserved word `property` documents that definition's intention.
-* Also Cryptol's `:check` with check all functions marked as properties in one go.
+* We can go a step further and `:prove` this property:
+  ```sh
+  Main> :prove absNonnegative 
+  Using random testing.
+  Passed 100 tests.
+  ```
+* Also Cryptol's `:check` with check all functions marked as
+  properties in one go and, you guessed it, `:prove` works similarly.
```

Comment:
Whoa -- you're telling me that `:prove` doesn't actually **prove**?!?  Mind blown...

...Okay, my confusion mounts: later text speaks of `:prove` actually proving (or trying until the heat death of the universe).  So what's going on in this example where `:prove` invokes some number of tests?

Reply:
copy and past-o fixing.

Reply:
In this example, the prover never even sees the formula, because the recursive function isn't "symbolically" terminating. That is, Cryptol's symbolic execution engine can't tell when to stop recursing...so it trys to build an infinitely huge formula.

Reply:
Do we want to add this to the exposition?

Reply:
It now says: 
  ```sh
  labs::LanguageBasics::LanguageBasics> :prove absNonnegative 
  Q.E.D.
  (Total Elapsed Time: 0.032s, using Z3)
  ```